test/tools/cmap_spec_runner.ts

```diff
@@ -0,0 +1,336 @@
+import { expect } from 'chai';
+import { EventEmitter } from 'events';
+import * as util from 'util';
+
+import { Connection } from '../../src';
+import { ConnectionPool } from '../../src/cmap/connection_pool';
+import { sleep } from './utils';
+
+type cmapOperation =
+  | { name: 'start' | 'waitForThread'; target: string }
+  | { name: 'wait'; ms: number }
+  | { name: 'waitForEvent'; event: string; count: number }
+  | { name: 'checkOut'; thread: string; label: string }
+  | { name: 'checkIn'; connection: string }
+  | { name: 'clear' | 'close' | 'ready' };
+
+type cmapPoolOptions = {
+  maxPoolSize?: number;
+  minPoolSize?: number;
+  maxIdleTimeMS?: number;
+  waitQueueTimeoutMS?: number;
+};
+
+type cmapEvent = {
+  type: string;
+  address?: 42;
+  connectionId?: number;
+  options?: 42 | cmapPoolOptions;
+  reason: string;
+};
+
+const knownTestKeys = [
+  'name',
+  'version',
+  'style',
+  'description',
+  'poolOptions',
+  'operations',
+  'error',
+  'events',
+  'ignore'
+];
+export type cmapTest = {
+  name?: string; // filename path added by the spec loader
```

Comment:
If it's always added is it optional?

Reply:
it's only added if you load the tests via the spec loader, it's conceivable that you may parse the definitions directly and this should still work, since it's not actually part of the json test definition